File: pacengine/engine/trace/src/TraceReader.cpp
#include "TraceReader.hpp"

#include "TraceFormat.hpp"

#include <cstring>
#include <fstream>
#include <sstream>
#include <vector>

namespace pac {

namespace {

class Cursor {
public:
    Cursor(const std::uint8_t* data, std::size_t size)
        : data_(data), size_(size) {}

    bool         eof() const noexcept { return pos_ >= size_; }
    std::size_t  pos() const noexcept { return pos_; }
    bool         good() const noexcept { return ok_; }

    bool need(std::size_t n) {
        if (pos_ + n > size_) {
            ok_ = false;
            return false;
        }
        return true;
    }

    std::uint8_t read_u8() {
        if (!need(1)) return 0;
        return data_[pos_++];
    }

    std::uint16_t read_u16() {
        if (!need(2)) return 0;
        const std::uint16_t v = static_cast<std::uint16_t>(data_[pos_]) |
                                (static_cast<std::uint16_t>(data_[pos_ + 1]) << 8);
        pos_ += 2;
        return v;
    }

    std::uint32_t read_u32() {
        if (!need(4)) return 0;
        const std::uint32_t v =  static_cast<std::uint32_t>(data_[pos_]) |
                                (static_cast<std::uint32_t>(data_[pos_ + 1]) << 8) |
                                (static_cast<std::uint32_t>(data_[pos_ + 2]) << 16) |
                                (static_cast<std::uint32_t>(data_[pos_ + 3]) << 24);
        pos_ += 4;
        return v;
    }

    std::uint64_t read_u64() {
        if (!need(8)) return 0;
        std::uint64_t v = 0;
        for (int i = 0; i < 8; ++i) {
            v |= static_cast<std::uint64_t>(data_[pos_ + i]) << (8 * i);
        }
        pos_ += 8;
        return v;
    }

    std::vector<std::uint8_t> read_bytes(std::size_t n) {
        if (!need(n)) return {};
        std::vector<std::uint8_t> out(data_ + pos_, data_ + pos_ + n);
        pos_ += n;
        return out;
    }

    std::string read_string_u32() {
        const std::uint32_t n = read_u32();
        if (!need(n)) return {};
        std::string s(reinterpret_cast<const char*>(data_ + pos_), n);
        pos_ += n;
        return s;
    }

private:
    const std::uint8_t* data_;
    std::size_t         size_;
    std::size_t         pos_ = 0;
    bool                ok_  = true;
};

} // namespace

TraceReader::TraceReader(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) {
        error_ = "TraceReader: cannot open " + path;
        return;
    }
    std::ostringstream buf;
    buf << in.rdbuf();
    const std::string  raw    = buf.str();
    const std::uint8_t* data  = reinterpret_cast<const std::uint8_t*>(raw.data());
    Cursor cur(data, raw.size());

    if (!cur.need(trace_format::kHeaderSize)) {
        error_ = "TraceReader: file shorter than v2 header";
        return;
    }
    if (std::memcmp(data, trace_format::kMagic, 4) != 0) {
        error_ = "TraceReader: bad magic (expected PACT)";
        return;
    }
    cur.read_bytes(4); // magic
    version_ = cur.read_u16();
    cur.read_u16(); // flags
    cur.read_u64(); // reserved

    if (version_ != trace_format::kVersionV2) {
        std::ostringstream e;
        e << "TraceReader: unsupported trace version " << version_;
        error_ = e.str();
        return;
    }

    while (!cur.eof()) {
        const std::uint32_t frame_size = cur.read_u32();
        if (!cur.good()) {
            error_ = "TraceReader: truncated frame size";
            return;
        }
        const std::size_t frame_start = cur.pos();
        if (!cur.need(frame_size)) {
            error_ = "TraceReader: frame size exceeds file";
            return;
        }

        TraceFrame frame;
        frame.tick = cur.read_u64();
        const std::uint32_t entity_count = cur.read_u32();
        frame.entities.reserve(entity_count);
        for (std::uint32_t i = 0; i < entity_count && cur.good(); ++i) {
            TraceEntity ent;
            ent.index      = cur.read_u32();
            ent.generation = cur.read_u32();
            const std::uint8_t comp_count = cur.read_u8();
            ent.components.reserve(comp_count);
            for (std::uint8_t c = 0; c < comp_count && cur.good(); ++c) {
                TraceComponent comp;
                comp.type_tag = cur.read_u16();
                const std::uint16_t payload_size = cur.read_u16();
                comp.payload  = cur.read_bytes(payload_size);
                ent.components.push_back(std::move(comp));
            }
            frame.entities.push_back(std::move(ent));
        }
        const std::uint32_t event_count = cur.read_u32();
        frame.events.reserve(event_count);
        for (std::uint32_t i = 0; i < event_count && cur.good(); ++i) {
            frame.events.push_back(cur.read_string_u32());
        }
        if (!cur.good()) {
            error_ = "TraceReader: truncated frame body";
            return;
        }
        if (cur.pos() != frame_start + frame_size) {
            error_ = "TraceReader: frame size mismatch (parser drift)";
            return;
        }
        frames_.push_back(std::move(frame));
    }

    ok_ = true;
}

const TraceFrame* TraceReader::frame_at_tick(std::uint64_t tick) const {
    // Frames are written in monotonically increasing tick order.
    // Linear scan is fine for v1; switch to binary search later.
    for (const auto& f : frames_) {
        if (f.tick == tick) return &f;
    }
    return nullptr;
}

std::optional<std::string>
TraceReader::decode_string(const TraceComponent& c) {
    if (c.payload.size() < 2) return std::nullopt;
    const std::uint16_t n =
        static_cast<std::uint16_t>(c.payload[0]) |
        (static_cast<std::uint16_t>(c.payload[1]) << 8);
    if (c.payload.size() < std::size_t(2 + n)) return std::nullopt;
    return std::string(reinterpret_cast<const char*>(c.payload.data() + 2), n);
}

std::optional<std::tuple<double, double, double>>
TraceReader::decode_position(const TraceComponent& c) {
    if (c.type_tag != trace_format::kPositionTag) return std::nullopt;
    if (c.payload.size() != 24) return std::nullopt;
    auto read_f64 = [&](std::size_t off) {
        std::uint64_t bits = 0;
        for (int i = 0; i < 8; ++i) {
            bits |= static_cast<std::uint64_t>(c.payload[off + i]) << (8 * i);
        }
        double d;
        std::memcpy(&d, &bits, sizeof(d));
        return d;
    };
    return std::make_tuple(read_f64(0), read_f64(8), read_f64(16));
}

} // namespace pac

File: artifacts/api-server/src/lib/trace-reader.ts
/*
 * Trace v2 reader — TS mirror of pacengine/engine/trace/src/TraceReader.cpp.
 *
 * The exact same byte format that the C++ writer produces is parsed
 * here. Determinism note: the editor never re-encodes a trace; this
 * reader only decodes it into JSON-friendly shapes for the
 * /pacengine/runs/:runId/frames endpoint.
 */

import type { TraceFrame } from "@workspace/api-zod/types/traceFrame";
import type { EntityFrame } from "@workspace/api-zod/types/entityFrame";
import type { PositionValue } from "@workspace/api-zod/types/positionValue";

export type { TraceFrame, EntityFrame, PositionValue };

// Component type tags — must match TraceFormat.hpp.
const TAG_PAC_ID = 1;
const TAG_ENTITY_TYPE = 2;
const TAG_POSITION = 3;

const HEADER_SIZE = 16;
const MAGIC = "PACT";
const VERSION_V2 = 2;

class Cursor {
  pos = 0;
  constructor(private readonly buf: Buffer) {}
  remaining(): number {
    return this.buf.length - this.pos;
  }
  readU8(): number {
    const v = this.buf.readUInt8(this.pos);
    this.pos += 1;
    return v;
  }
  readU16(): number {
    const v = this.buf.readUInt16LE(this.pos);
    this.pos += 2;
    return v;
  }
  readU32(): number {
    const v = this.buf.readUInt32LE(this.pos);
    this.pos += 4;
    return v;
  }
  readU64(): bigint {
    const v = this.buf.readBigUInt64LE(this.pos);
    this.pos += 8;
    return v;
  }
  readF64(): number {
    const v = this.buf.readDoubleLE(this.pos);
    this.pos += 8;
    return v;
  }
  readBytes(n: number): Buffer {
    const out = this.buf.subarray(this.pos, this.pos + n);
    this.pos += n;
    return out;
  }
}

export interface ParsedTrace {
  version: number;
  frames: TraceFrame[];
}

export class TraceParseError extends Error {}

export function parseTraceV2(buf: Buffer): ParsedTrace {
  if (buf.length < HEADER_SIZE) {
    throw new TraceParseError("trace file shorter than 16-byte header");
  }
  if (buf.subarray(0, 4).toString("ascii") !== MAGIC) {
    throw new TraceParseError("bad magic (expected 'PACT')");
  }
  const version = buf.readUInt16LE(4);
  if (version !== VERSION_V2) {
    throw new TraceParseError(`unsupported trace version ${version}`);
  }

  const cur = new Cursor(buf);
  cur.pos = HEADER_SIZE;

  const frames: TraceFrame[] = [];
  while (cur.remaining() > 0) {
    if (cur.remaining() < 4) {
      throw new TraceParseError("truncated frame size");
    }
    const frameSize = cur.readU32();
    if (cur.remaining() < frameSize) {
      throw new TraceParseError("frame size exceeds file");
    }
    const frameStart = cur.pos;

    const tickBig = cur.readU64();
    // Ticks fit in 53-bit JS number for any realistic editor session.
    const tick = Number(tickBig);

    const entityCount = cur.readU32();
    const entities: EntityFrame[] = [];
    for (let i = 0; i < entityCount; i++) {
      const index = cur.readU32();
      const generation = cur.readU32();
      const compCount = cur.readU8();

      const ef: EntityFrame = { index, generation };
      for (let c = 0; c < compCount; c++) {
        const tag = cur.readU16();
        const payloadSize = cur.readU16();
        const payload = cur.readBytes(payloadSize);
        switch (tag) {
          case TAG_PAC_ID: {
            ef.pacId = readStringU16(payload);
            break;
          }
          case TAG_ENTITY_TYPE: {
            ef.type = readStringU16(payload);
            break;
          }
          case TAG_POSITION: {
            if (payload.length === 24) {
              const pos: PositionValue = {
                x: payload.readDoubleLE(0),
                y: payload.readDoubleLE(8),
                z: payload.readDoubleLE(16),
              };
              ef.position = pos;
            }
            break;
          }
          default:
            // Forward-compat: unknown component types are skipped silently.
            break;
        }
      }
      entities.push(ef);
    }

    const eventCount = cur.readU32();
    const events: string[] = [];
    for (let e = 0; e < eventCount; e++) {
      const n = cur.readU32();
      events.push(cur.readBytes(n).toString("utf8"));
    }

    if (cur.pos !== frameStart + frameSize) {
      throw new TraceParseError(
        `frame size mismatch (parser drift) at frame ${frames.length}`,
      );
    }
    frames.push({ tick, entities, events });
  }

  return { version, frames };
}

function readStringU16(payload: Buffer): string {
  if (payload.length < 2) return "";
  const n = payload.readUInt16LE(0);
  if (payload.length < 2 + n) return "";
  return payload.subarray(2, 2 + n).toString("utf8");
}
